### PR TITLE
Use zero copy to send file if can

### PR DESCRIPTION
**Pull request by abersheeran**
https://asgi.readthedocs.io/en/latest/extensions.html#zero-copy-send

### CODE REVIEW

File: starlette/responses.py [hunk 1]
```diff
@@ -296,6 +296,22 @@ async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
...
+            try:
+                await send(
+                    {
+                        "type": "http.response.zerocopysend",
+                        "file": fd,
+                        "more_body": False,
+                    }
+                )
+            finally:
+                os.close(fd)
```
Comment (Kludex) [inline starlette/responses.py]:
```suggestion
                await anyio.to_thread.run_sync(os.close, fd)
```

File: starlette/responses.py [hunk 2]
```diff
@@ -296,6 +296,22 @@ async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
         )
         if self.send_header_only:
             await send({"type": "http.response.body", "body": b"", "more_body": False})
+        elif "http.response.zerocopysend" in scope["extensions"]:  # pragma: no cover
+            if os.name == "nt":
+                open_mode = os.O_RDONLY | os.O_BINARY
+            else:
+                open_mode = os.O_RDONLY
+            fd = os.open(self.path, open_mode)
```
Comment (Kludex) [inline starlette/responses.py]:
```suggestion
            fd = await anyio.to_thread.run_sync(os.open, self.path, open_mode)
```

**Review by Kludex — CHANGES_REQUESTED** [review]

File: starlette/responses.py [hunk 3]
```diff
@@ -313,6 +313,22 @@ async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
         )
         if self.send_header_only:
             await send({"type": "http.response.body", "body": b"", "more_body": False})
+        elif "http.response.zerocopysend" in scope["extensions"]:  # pragma: no cover
```
Comment (Kludex) [inline starlette/responses.py:330]: Why should we not test it?

**Comment by synodriver** [comment]
By the way, should we use [range](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range) and [if-range](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Range) header? I saw it in [someones toolkit](https://github.com/abersheeran/baize/blob/5e84d69c2389aa46c81e5f4e96de54db4f0f1d3b/baize/asgi.py#L630-L640)

File: starlette/responses.py (hunk 3 above)
Comment (abersheeran) [reply starlette/responses.py:330]: I'm not sure how to test it. It seems too simple to just test if the result it returns is a dictionary. Do you have any suggestions?

**Comment by abersheeran** [comment]
> By the way, should we use [range](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range) and [if-range](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Range) header? I saw it in [someones toolkit](https://github.com/abersheeran/baize/blob/5e84d69c2389aa46c81e5f4e96de54db4f0f1d3b/baize/asgi.py#L630-L640)

https://github.com/encode/starlette/pull/1013

**Comment by synodriver** [comment]
> > By the way, should we use [range](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range) and [if-range](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Range) header? I saw it in [someones toolkit](https://github.com/abersheeran/baize/blob/5e84d69c2389aa46c81e5f4e96de54db4f0f1d3b/baize/asgi.py#L630-L640)
> 
> #1013

Than you for pointing out. But is that means ```zerocopysend``` have to do something in order to implement that? Like adding ```offset``` and ```count``` params.

**Comment by abersheeran** [comment]
In addition, `baize.asgi.FileResponse` can work with Starlette. Just try the following code

```python
# main.py
from starlette.applications import Starlette
from starlette.routing import Route
from baize.asgi import FileResponse


async def homepage(request):
    return FileResponse("./main.py")

routes = [
    Route("/", endpoint=homepage)
]

app = Starlette(debug=True, routes=routes)
```

**Comment by abersheeran** [comment]
> > > By the way, should we use [range](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range) and [if-range](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Range) header? I saw it in [someones toolkit](https://github.com/abersheeran/baize/blob/5e84d69c2389aa46c81e5f4e96de54db4f0f1d3b/baize/asgi.py#L630-L640)
> > 
> > 
> > #1013
> 
> Than you for pointing out. But is that means `zerocopysend` have to do something in order to implement that? Like adding `offset` and `count` params.

Yes. These parameters need to be specified.

**Comment by synodriver** [comment]
> In addition, `baize.asgi.FileResponse` can work with Starlette. Just try the following code
> 
> ```python
> # main.py
> from starlette.applications import Starlette
> from starlette.routing import Route
> from baize.asgi import FileResponse
> 
> 
> async def homepage(request):
>     return FileResponse("./main.py")
> 
> routes = [
>     Route("/", endpoint=homepage)
> ]
> 
> app = Starlette(debug=True, routes=routes)
> ```

Great feature!

**Comment by abersheeran** [comment]
> > In addition, `baize.asgi.FileResponse` can work with Starlette. Just try the following code
> > ```python
> > # main.py
> > from starlette.applications import Starlette
> > from starlette.routing import Route
> > from baize.asgi import FileResponse
> > 
> > 
> > async def homepage(request):
> >     return FileResponse("./main.py")
> > 
> > routes = [
> >     Route("/", endpoint=homepage)
> > ]
> > 
> > app = Starlette(debug=True, routes=routes)
> > ```
> 
> Great feature!

Simply because they support the ASGI3 protocol. I was fascinated by this design in Starlette, so I implemented a similar part for WSGI.

**Review by synodriver — APPROVED** [review]

File: starlette/responses.py [hunk 4]
```diff
@@ -327,6 +327,22 @@ async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
...
+            try:
+                await send(
+                    {
+                        "type": "http.response.zerocopysend",
+                        "file": fd,
+                        "more_body": False,
+                    }
+                )
+            finally:
+                await anyio.to_thread.run_sync(os.close, fd)
```
Comment (lovelydinosaur) [inline starlette/responses.py:345]: I'm... not comfortable with this.
Not at all obvious to me how the open/zerocopesend/close would interact.

I've got a bit of a preference here that we close this off for now as a "feature pause" until we're sufficiently on top of all existing issues that we can *thoroughly* review this.

**Pull request closed without merging**